rfcs/text/0032-structured-benchmark-programs.md

```diff
@@ -0,0 +1,192 @@
+- *Block Encoding*: Block encoding embeds classical data or operators into quantum states. It underpins advanced algorithms like QSVT. It utilizes conditionals to apply different operations. ([Quantum Singular Value Transformation (QSVT)](https://arxiv.org/abs/1806.01838), [Low & Chuang, 2016](https://arxiv.org/abs/1606.02685))
+
+## Mixed Control Flow
+
+*This class of benchmark programs combines loops and conditionals to implement more complex algorithms or protocols.*
+
+- *Quantum Multiplexers*: Quantum multiplexers apply different unitary operations based on the state of control qubits, effectively implementing conditionals in quantum circuits. ([Quantum multiplexers in circuit synthesis](https://arxiv.org/abs/quant-ph/0410066))
+
+## Dynamic Qubit (Re-)Allocation
+
+*This class of benchmark programs allocates qubits dynamically during execution, based on runtime conditions or within loop bodies.*
+
+- *Iterative Quantum Fourier Transform (iQFT)*: Rather than allocating all qubits upfront, just a single qubit is allocated at the start of execution and then reset during each loop iteration of QFT.
+- *Iterative Quantum Phase Estimation (iQPE)*: Similar to iQFT, iQPE allocates a single qubit (in addition to the phase qubit "psi") and reuses it across iterations to estimate the phase of a unitary operator.
+- *Shor's Algorithm*: The primitives used in Shor's algorithm can similarly be implemented using dynamic qubit reallocation to minimize the total number of qubits required.
+- *Toffoli-heavy Circuits*: Toffoli gates in quantum circuits can be implemented more efficiently using ancilla qubits. Circuits using a larger number of Toffoli gates, such as many arithmetic circuits, can benefit from dynamic qubit allocation to allocate and deallocate ancilla qubits repeatedly inside loop bodies.
+
+## Loosely Coupled Hybrid Programs
+
+*Variational and hybrid algorithms naturally combine quantum circuits with classical optimization loops. They are loosely coupled because the quantum and classical parts interact only at well-defined synchronization points.*
+
+- *VQE*: A variational quantum eigensolver algorithm that uses a classical optimizer to minimize the energy of a parameterized quantum circuit which implements several iterations of some ansatz. ([Peruzzo et al., 2014](https://arxiv.org/abs/1304.3061))
+- *QAOA*: A quantum approximate optimization algorithm that uses a classical optimizer to find optimal parameters for a parameterized quantum circuit. It iteratively applies a "problem" and "mixer" operator using loop structures. ([Farhi et al., 2014](https://arxiv.org/abs/1411.4028))
+
+## Fault-Tolerant State Preparation and Error Correction
+
+*Quantum Error Correction (QEC) is one of the most important applications of structured control flow in quantum computing. These benchmark programs implement QEC protocols that involve structured operations at any point in the program.*
```

Comment:
As I've gone down the list, I wonder if it's worth making a distinction between "simple" benchmark that utilize a single structured primitive, and "compound" benchmarks that combine multiple primitives.

For example, QFT is a simple benchmark that uses a dynamically-sized loop that indexes qubits. QPE is compound in the sense that it does this in two places (controlled rotations, plus the QFT). Something like error correction is really complex as it weaves many primitives together.